rllib/agents/alpha_star/league_builder.py

```diff
@@ -0,0 +1,356 @@
+
+        # Train all non-random policies that exist at beginning.
+        ma_config["policies_to_train"] = ["main_0"]
+
+        # Add random policies.
+        i = -1
+        for i in range(num_random_policies // 2):
+            policies[f"league_exploiter_{i}"] = PolicySpec(policy_class=RandomPolicy)
+            policies[f"main_exploiter_{i}"] = PolicySpec(policy_class=RandomPolicy)
+        # Add initial (learning) league-exploiters.
+        for j in range(num_learning_league_exploiters):
+            pid = f"league_exploiter_{j + i + 1}"
+            policies[pid] = PolicySpec()
+            ma_config["policies_to_train"].append(pid)
+        # Add initial (learning) main-exploiters.
+        for j in range(num_learning_league_exploiters):
+            pid = f"main_exploiter_{j + i + 1}"
+            policies[pid] = PolicySpec()
+            ma_config["policies_to_train"].append(pid)
+
+        # Initial policy mapping function: main_0 vs main_exploiter_0.
+        ma_config["policy_mapping_fn"] = (
+            lambda aid, ep, worker, **kw: "main_0"
+            if ep.episode_id % 2 == aid
+            else "main_exploiter_0"
+        )
+
+    @override(LeagueBuilder)
+    def build_league(self, result: ResultDict) -> None:
+        local_worker = self.trainer.workers.local_worker()
+
+        # If no evaluation results -> Use hist data gathered for training.
+        if "evaluation" in result:
+            hist_stats = result["evaluation"]["hist_stats"]
+        else:
+            hist_stats = result["hist_stats"]
+
+        trainable_policies = local_worker.get_policies_to_train()
+        non_trainable_policies = (
+            set(local_worker.policy_map.keys()) - trainable_policies
+        )
+
+        logger.info(f"League building after iter {self.trainer.iteration}:")
+
+        # Calculate current win-rates.
+        for policy_id, rew in hist_stats.items():
+            mo = re.match("^policy_(.+)_reward$", policy_id)
+            if mo is None:
+                continue
+            policy_id = mo.group(1)
+
+            # Calculate this policy's win rate.
+            won = 0
+            for r in rew:
+                if r > 0.0:  # win = 1.0; loss = -1.0
+                    won += 1
+            win_rate = won / len(rew)
+            # TODO: This should probably be a running average
+            #  (instead of hard-overriding it with the most recent data).
+            self.trainer.win_rates[policy_id] = win_rate
+
+            # Policy is a snapshot (frozen) -> Ignore.
+            if policy_id not in trainable_policies:
+                continue
+
+            logger.info(f"\t{policy_id} win-rate={win_rate} -> ", end="")
+
+            # If win rate is good enough -> Snapshot current policy and decide,
+            # whether to freeze the new snapshot or not.
+            if win_rate >= self.win_rate_threshold_for_new_snapshot:
+                is_main = re.match("^main(_\\d+)?$", policy_id)
```

Comment:
do we really need re, or we can simply policy_id.starts_with(...)?

Reply:
If we use "startswith", then it could be a `main_exploiter_\\d` as well.
I like `re` (and Perl!) very much :)